mermithid/misc/ComplexLineShapeUtilities.py

```diff
@@ -0,0 +1,192 @@
+import numpy as np
+import json
+from mermithid.misc import Constants
+
+# Natural constants
+kr_line = Constants.kr_line() # 17.8260 keV
+kr_line_width = Constants.kr_line_width() # 2.83 eV
+e_charge = Constants.e_charge() # 1.60217662*10**(-19) Coulombs , charge of electron
+m_e = Constants.m_e() # 9.10938356*10**(-31) Kilograms , mass of electron
+mass_energy_electron = Constants.m_electron()/1000 # 510.9989461 keV
+
+# A lorentzian function
+def lorentzian(x_array,x0,FWHM):
+    HWHM = FWHM/2.
+    func = (1./np.pi)*HWHM/((x_array-x0)**2.+HWHM**2.)
+    return func
+
+# A gaussian function
+def gaussian(x_array,A,sigma,mu):
+    f = A*(1./(sigma*np.sqrt(2*np.pi)))*np.exp(-(((x_array-mu)/sigma)**2.)/2.)
+    return f
+
+# Converts a gaussian's FWHM to sigma
+def gaussian_FWHM_to_sigma(FWHM):
+    sigma = FWHM/(2*np.sqrt(2*np.log(2)))
+    return sigma
+
+# Converts a gaussian's sigma to FWHM
+def gaussian_sigma_to_FWHM(sigma):
+    FWHM = sigma*(2*np.sqrt(2*np.log(2)))
+    return FWHM
+
+#returns array with energy loss/ oscillator strength data
+def read_oscillator_str_file(filename):
+    f = open(filename, "r")
+    lines = f.readlines()
+    f.close()
+    energyOsc = [[],[]] #2d array of energy losses, oscillator strengths
+
+    for line in lines:
+        if line != "" and line[0]!="#":
+            raw_data = [float(i) for i in line.split("\t")]
+            energyOsc[0].append(raw_data[0])
+            energyOsc[1].append(raw_data[1])
+
+    energyOsc = np.array(energyOsc)
+    ### take data and sort by energy
+    sorted_indices = energyOsc[0].argsort()
+    energyOsc[0] = energyOsc[0][sorted_indices]
+    energyOsc[1] = energyOsc[1][sorted_indices]
+    return energyOsc
+
+# A sub function for the scatter function. Found in
+# "Energy loss of 18 keV electrons in gaseous T and quench condensed D films"
+# by V.N. Aseev et al. 2000
+def aseev_func_tail(energy_loss_array, gas_type):
+    if gas_type=="H2":
+        A2, omeg2, eps2 = 0.195, 14.13, 10.60
+    elif gas_type=="Kr":
+        A2, omeg2, eps2 = 0.4019, 22.31, 16.725
+    return A2*omeg2**2./(omeg2**2.+4*(energy_loss_array-eps2)**2.)
+
+#convert oscillator strength into energy loss spectrum
+def get_eloss_spec(e_loss, oscillator_strength, kr_line): #energies in eV
+    kinetic_en = kr_line * 1000
+    e_rydberg = 13.605693009 #rydberg energy (eV)
+    a0 = 5.291772e-11 #bohr radius
+    return np.where(e_loss>0 , 4.*np.pi*a0**2 * e_rydberg / (kinetic_en * e_loss) * oscillator_strength * np.log(4. * kinetic_en * e_loss / (e_rydberg**3.) ), 0)
+
+# Takes only the nonzero bins of a histogram
+def get_only_nonzero_bins(bins,hist):
```

Comment:
Careful here! you are using binned data to compute error bars.
Since you are later on using curve_fit which does a least square technique, you will misestimate this error: for testing purposes, this is "fine"; for papers material, this should be corrected in my opinion.
You should use the model value in that bin (and not the measured content of the bin) to do so, otherwise you will have issues as described by Hamish in a recent Basecamp post...
Several possible options:
- keep using curve_fit but using the model for computing the error
- use a binned (aka Poisson) Likelihood minimization technique (not Chi2) and use Cramer-Rao inequality for computing the errors
- use RooFit (like the https://github.com/morphoorg/morpho/blob/master/morpho/processors/sampling/PyBindRooFitProcessor.py ) that already knows how to deal with data errors for histograms